04-onion-routing.md

```diff
@@ -855,76 +883,66 @@ func NewOnionPacket(paymentPath []*btcec.PublicKey, sessionKey *btcec.PrivateKey
-time-constant to avoid information leaks.
-
-At this point, the processing node can generate a _rho_-key.
-
-The routing information is then deobfuscated, and the information about the
-next hop is extracted.
-To do so, the processing node copies the `hop_payloads` field, appends 1300 `0x00`-bytes,
-generates `2*1300` pseudo-random bytes (using the _rho_-key), and applies the result, using `XOR`, to the copy of the `hop_payloads`.
-The first few bytes correspond to the bigsize-encoded length `l` of the `hop_payload`, followed by `l` bytes of the resulting routing information become the `hop_payload`, and the 32 byte HMAC.
-The next 1300 bytes are the `hop_payloads` for the outgoing packet.
-
-A special `hmac` value of 32 `0x00`-bytes indicates that the currently processing hop is the intended recipient and that the packet should not be forwarded.
-
-If the HMAC does not indicate route termination, and if the next hop is a peer of the
-processing node; then the new packet is assembled. Packet assembly is accomplished
-by blinding the ephemeral key with the processing node's public key, along with the
-shared secret, and by serializing the `hop_payloads`.
-The resulting packet is then forwarded to the addressed peer.
+Those sections specify the `associated_data` and `blinding` to use, the extracted payload format and handling (including how to determine the next peer, if any), and how to handle errors.  The processing itself is identical.
 
 ## Requirements
 
-The processing node:
-  - if the ephemeral public key is NOT on the `secp256k1` curve:
-    - MUST abort processing the packet.
-    - MUST report a route failure to the origin node.
-  - if the packet has previously been forwarded or locally redeemed, i.e. the
-  packet contains duplicate routing information to a previously received packet:
-    - if preimage is known:
+A reader:
+  - if `version` is not 0:
+    - MUST abort processing the packet and fail.
+  - if `public_key` is not a valid pubkey:
+    - MUST abort processing the packet and fail.
+  - if the onion is for a payment:
+  - if `hmac` has previously been received:
+    - if the preimage is known:
       - MAY immediately redeem the HTLC using the preimage.
     - otherwise:
-      - MUST abort processing and report a route failure.
-  - if the computed HMAC and the packet's HMAC differ:
-    - MUST abort processing.
-    - MUST report a route failure.
-  - if the `realm` is unknown:
-    - MUST drop the packet.
-    - MUST signal a route failure.
-  - MUST address the packet to another peer that is its direct neighbor.
-  - if the processing node does not have a peer with the matching address:
-    - MUST drop the packet.
-    - MUST signal a route failure.
+      - MUST abort processing the packet and fail.
+  - if `blinding` is specified:
+    - Calculate the `blinding_ss` as ECDH(`blinding`, `node-privkey`)
+    - Tweak `public_key` by multiplying by $`HMAC256(\text{"blinded\_node\_id}", blinding_ss)`$
```

Comment:
This is confusing. We need to tweak the **private** key. And then we need to use the tweaked key instead of the public one.

Reply:
This is the alternative construction, where we tweak the onion ephemeral key (which is equivalent).  (CLN needs to do this because we have an HSM and we didn't want to teach it to tweak, but it's also easier in the text).

We could specify both, but we do note the private key method alternative in the rationale?